# NetRNApan
## Deciphering RNA modification and post-transcriptional regulation by deep learning framework 

RNA modification, which is evolutionarily conserved, is crucial for modulating various biological functions and disease pathogenesis. High resolution transcriptome-wide mapping of RNA modifications has facilitated both data resources and computational prediction of RNA modification. While these prediction algorithms are promising, they are limited in interpretability or generalizability, or the capacity for discovering novel post-transcriptional regulations. Here, we present NetRNApan, a deep learning framework for RNA modification site prediction, motif discovery and trans-regulatory factor identification. Using m5U profiles generated by FICC-seq and miCLIP-seq technologies as cases, we demonstrated the accuracy of NetRNApan with more efficient and interpretive feature representations. Decoding the informative characteristics detected by NetRNApan uncovered five representative clusters with consensus motifs that may be essential for m5U modification. Furthermore, NetRNApan revealed interesting trans-regulatory factors and provided a protein-binding perspective for investigating the function of RNA modifications. Specifically, we discovered 21 potential functional RNA-binding proteins (RBPs) whose binding sites were significantly linked to the extracted top-scoring motifs. Two examples are ANKHD1 and RBM4 with potential regulatory function of RNA modifications. Further analyses of identified RBPs revealed new insights into post-transcriptional regulatory mechanisms of m5U, such as gene expression, RNA splicing, and RNA transport. NetRNApan and the findings will be helpful for accurate and high-throughput detection of RNA modification sites in the study of mRNA regulation. NetRNApan is freely available at https://github.com/bsml320/NetRNApan.

<div align=center><img src="https://bioinfo.uth.edu/iapp/github/Figure1.jpg" width="800px"></div>

# Installation
Download ImmuneApp by
```
git clone https://github.com/bsml320/ImmuneApp
```
Installation has been tested in Linux server, CentOS Linux release 7.8.2003 (Core), with Python 3.7. Since the package is written in python 3x, python3x with the pip tool must be installed. ImmuneApp uses the following dependencies: numpy, scipy, pandas, h5py, keras version=2.3.1, tensorflow=1.15, seaborn, logomaker, and shutil, pathlib. We highly recommend that users leave a message under the ImmuneApp issue interface (https://github.com/bsml320/ImmuneApp/issues) when encountering any installation and running problems. We will deal with it in time. You can install these packages by the following commands:
```
conda create -n ImmuneApp python=3.7
conda activate ImmuneApp
pip install pandas
pip install numpy
pip install scipy
pip install -v keras==2.3.1
pip install -v tensorflow==1.15
pip install seaborn
pip install logomaker
pip install shutil
pip install pathlib
pip install protobuf==3.20
pip install h5py==2.10.0
```
__Note: please unzip the gibbscluster-2.0f.Linux.tar.gz in the folder first, and modify the "FULL PATH" in the gibbscluster execution file__

# Usage
### ImmuneApp provides two services: precise antigen presentation prediction and clinical immunopeptidomic cohorts analysis.

__1__. For antigen presentation prediction, this module accept two different types of input; FASTA and Peptide. In addition, candidate HLA molecules should be specified in the predictions. For FASTA input, the peptide length(s) should be specified.
