# NetRNApan
## Deciphering RNA modification and post-transcriptional regulation by deep learning framework 

RNA modification, which is evolutionarily conserved, is crucial for modulating various biological functions and disease pathogenesis. High resolution transcriptome-wide mapping of RNA modifications has facilitated both data resources and computational prediction of RNA modification. While these prediction algorithms are promising, they are limited in interpretability or generalizability, or the capacity for discovering novel post-transcriptional regulations. Here, we present NetRNApan, a deep learning framework for RNA modification site prediction, motif discovery and trans-regulatory factor identification. Using m5U profiles generated by FICC-seq and miCLIP-seq technologies as cases, we demonstrated the accuracy of NetRNApan with more efficient and interpretive feature representations. Decoding the informative characteristics detected by NetRNApan uncovered five representative clusters with consensus motifs that may be essential for m5U modification. Furthermore, NetRNApan revealed interesting trans-regulatory factors and provided a protein-binding perspective for investigating the function of RNA modifications. Specifically, we discovered 21 potential functional RNA-binding proteins (RBPs) whose binding sites were significantly linked to the extracted top-scoring motifs. Two examples are ANKHD1 and RBM4 with potential regulatory function of RNA modifications. Further analyses of identified RBPs revealed new insights into post-transcriptional regulatory mechanisms of m5U, such as gene expression, RNA splicing, and RNA transport. NetRNApan and the findings will be helpful for accurate and high-throughput detection of RNA modification sites in the study of mRNA regulation. NetRNApan is freely available at https://github.com/bsml320/NetRNApan.

<div align=center><img src="https://bioinfo.uth.edu/iapp/github/NetRNApan/Figure 1.jpg" width="800px"></div>

# Installation
Download NetRNApan by
```
git clone https://github.com/bsml320/NetRNApan
```
Installation has been tested in Linux server, CentOS Linux release 7.8.2003 (Core), with Python 3.7. Since the package is written in python 3x, python3x with the pip tool must be installed. NetRNApan uses the following dependencies: numpy, scipy, pandas, h5py, keras version=2.3.1, tensorflow=1.15 shutil, and pathlib. We highly recommend that users leave a message under the NetRNApan issue interface (https://github.com/bsml320/NetRNApan/issue) when encountering any installation and running problems. We will deal with it in time. You can install these packages by the following commands:
```
conda create -n NetRNApan python=3.7
conda activate NetRNApan
pip install pandas
pip install numpy
pip install scipy
pip install -v keras==2.3.1
pip install -v tensorflow==1.15
pip install seaborn
pip install shutil
pip install protobuf==3.20
pip install h5py==2.10.0
```
# Performance
First, the 10-fold CV was executed to assess NetRNApan's prediction performance on the training dataset. The ROC curves were drawn, and the corresponding AUC values were calculated. NetRNApan performed well, with an average AUC value of 0.9733 by 10-fold CV, ranging from 0.9697 to 0.9809, suggesting its good and reliable predictive power. NetRNApan prediction robustness was evaluated as well using the area under the precision-recall (AUPR) curve. The PR curve represents the trade-off between the number of false positive predictions and the number of false negative predictions. NetRNApan achieved AUPR values ranging from 0.9741 to 0.9819 during the balanced sample with 10-fold CV training, indicating that our model had great potential in predicting true positive sites with the high precision. Furthermore, we examined the generalizability of our models using an independent dataset that was not part of the training set. NetRNApan had the AUC value and AUPR of 0.9872 and 0.9877, respectively. These good and consistent metrics between 10-fold CV and independent testing indicated the promising accuracy and robustness. In addition, in our comparison of NetRNApan with other existing tools using an independent dataset, NetRNApan had AUC value improvement by over 3.5% (0.987 vs 0.954) for the modification sites prediction when compared to m5UPred. For performance comparison and benchmarking, we encoded 12 widely used features in RNA modification predictions, including nine sequence-based features (ANF, Binary, CKSNAP, DNC, ENAC, Kmer, NAC, TNC and RCKmer) and three types of physicochemical properties-based features (EIIP, NCP and PseDNC). 84 conventional machine-learning predictors were developed based on 12 types of features with seven algorithms (SVM, RF, LR, AB, SGD, DT, KNN and GB). The average AUC value of individual predictor was calculated based on 10-fold CV. In comparison, we found that the performance of NetRNApan was better than other 84 conventional machine-learning predictors, resulting in the AUC value improvements from 3.66-26.94%.

<div align=center><img src="https://bioinfo.uth.edu/iapp/github/NetRNApan/performance.jpg" width="800px"></div>

# Interpretability
To elucidate the capability of hierarchical representation by NetBCE, we visualized the epitopes and non-epitopes using UMAP (Uniform Manifold Approximation and Projection) method based on the feature representation at varied network layers. We found the feature representation came to be more discriminative along the network layer hierarchy. More specifically, the feature representations for epitopes and non-epitopes sites were mixed at the input layer. As the model continues to train, epitopes and non-epitopes tend to occur in very distinct regions with efficient feature representation. 

<div align=center><img src="https://bioinfo.uth.edu/iapp/github/NetRNApan/Figure 3.jpg" width="800px"></div>

# Motif discovery
To elucidate the capability of hierarchical representation by NetBCE, we visualized the epitopes and non-epitopes using UMAP (Uniform Manifold Approximation and Projection) method based on the feature representation at varied network layers. We found the feature representation came to be more discriminative along the network layer hierarchy. More specifically, the feature representations for epitopes and non-epitopes sites were mixed at the input layer. As the model continues to train, epitopes and non-epitopes tend to occur in very distinct regions with efficient feature representation. 

<div align=center><img src="https://bioinfo.uth.edu/iapp/github/NetRNApan/Figure 5.jpg" width="800px"></div>

# Post-transcriptional regulation
To elucidate the capability of hierarchical representation by NetBCE, we visualized the epitopes and non-epitopes using UMAP (Uniform Manifold Approximation and Projection) method based on the feature representation at varied network layers. We found the feature representation came to be more discriminative along the network layer hierarchy. More specifically, the feature representations for epitopes and non-epitopes sites were mixed at the input layer. As the model continues to train, epitopes and non-epitopes tend to occur in very distinct regions with efficient feature representation. 

<div align=center><img src="https://bioinfo.uth.edu/iapp/github/NetRNApan/Figure 6.jpg" width="800px"></div>

# Usage
Please cd to the NetBCE/prediction/ folder which contains predict.py.
Example: 
```
cd NetBCE/prediction/
python NetBCE_prediction.py -f ../testdata/test.fasta -o ../result/test_result
```
For details of other parameters, run:
```
python NetBCE_prediction.py --help
```
